server/util/protofile/protofile.go

```diff
@@ -125,17 +123,90 @@ func (w *BufferedProtoWriter) WriteProtoToStream(ctx context.Context, msg proto.
 	return nil
 }
 
+type blobReadResult struct {
+	data []byte
+	err  error
+}
+
+type blobFuture chan blobReadResult
+
+type blobQueue struct {
+	blobstore      interfaces.Blobstore
+	streamID       string
+	maxConnections int
+	futures        []blobFuture
+	numPopped      int
+	done           bool
+}
+
+func newBlobQueue(blobstore interfaces.Blobstore, streamID string) *blobQueue {
+	return &blobQueue{
+		blobstore:      blobstore,
+		streamID:       streamID,
+		maxConnections: 16,
+		futures:        make([]blobFuture, 0),
+		numPopped:      0,
+		done:           false,
+	}
+}
+
+func newBlobFuture() blobFuture {
+	return make(blobFuture, 1)
+}
+
+func (q *blobQueue) pushNewFuture(ctx context.Context) {
+	future := newBlobFuture()
+	sequenceNumber := len(q.futures)
+	q.futures = append(q.futures, future)
+	go func() {
+		defer close(future)
+
+		tmpFilePath := chunkName(q.streamID, sequenceNumber)
+		data, err := q.blobstore.ReadBlob(ctx, tmpFilePath)
+		if err != nil {
+			future <- blobReadResult{
+				data: nil,
+				err:  err,
+			}
+			return
+		}
+
+		future <- blobReadResult{
+			data: data,
+			err:  nil,
+		}
+	}()
+}
+
+func (q *blobQueue) pop(ctx context.Context) ([]byte, error) {
+	if q.done {
+		return nil, status.FailedPreconditionError("Queue has been exhausted.")
```

Comment:
maybe this should be ResourceExhaustedError? You know better than me.

Reply:
That sounds better! Done